<a href="https://colab.research.google.com/github/HadeelNasasrah/Prediction-of-Product-Sales/blob/main/Project_1_Part_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***We will continue to work on our sales prediction project. The goal of this step is to help the retailer by using machine learning to make predictions about future sales based on the data provided.***

***For Part 5, we will go back to your original, uncleaned, sales prediction dataset with the goal of preventing data leakage.***

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [42]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.pipeline import make_pipeline
from sklearn import set_config
set_config(transform_output='pandas')
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OrdinalEncoder
# set the default output to pandas
from sklearn import set_config
set_config(transform_output='pandas')
from sklearn.linear_model import LinearRegression
# Import evaluation metrics
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV

In [35]:
# Define the custom functions for regressoin evaluation
def regression_metrics(y_true, y_pred, label='', verbose = True, output_dict=False):
  # Get metrics
  mae = mean_absolute_error(y_true, y_pred)
  mse = mean_squared_error(y_true, y_pred)
  rmse = mean_squared_error(y_true, y_pred, squared=False)
  r_squared = r2_score(y_true, y_pred)
  if verbose == True:
    # Print Result with Label and Header
    header = "-"*60
    print(header, f"Regression Metrics: {label}", header, sep='\n')
    print(f"- MAE = {mae:,.3f}")
    print(f"- MSE = {mse:,.3f}")
    print(f"- RMSE = {rmse:,.3f}")
    print(f"- R^2 = {r_squared:,.3f}")
  if output_dict == True:
      metrics = {'Label':label, 'MAE':mae,
                 'MSE':mse, 'RMSE':rmse, 'R^2':r_squared}
      return metrics

def evaluate_regression(reg, X_train, y_train, X_test, y_test, verbose = True,
                        output_frame=False):
  # Get predictions for training data
  y_train_pred = reg.predict(X_train)

  # Call the helper function to obtain regression metrics for training data
  results_train = regression_metrics(y_train, y_train_pred, verbose = verbose,
                                     output_dict=output_frame,
                                     label='Training Data')
  print()
  # Get predictions for test data
  y_test_pred = reg.predict(X_test)
  # Call the helper function to obtain regression metrics for test data
  results_test = regression_metrics(y_test, y_test_pred, verbose = verbose,
                                  output_dict=output_frame,
                                    label='Test Data' )

  # Store results in a dataframe if ouput_frame is True
  if output_frame:
    results_df = pd.DataFrame([results_train,results_test])
    # Set the label as the index
    results_df = results_df.set_index('Label')
    # Set index.name to none to get a cleaner looking result
    results_df.index.name=None
    # Return the dataframe
    return results_df.round(3)


In [3]:
# Load Ames Prep for ML

df = pd.read_csv('/content/drive/MyDrive/Data science&AI/CodingDojo/02-IntroML/Week05/Data/sales_predictions_2023.csv')
#df = df.set_index("PID")
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

***Before we split data, we will drop duplicates and fix inconsistencies in categorical data***


***1.1 We will check the data types and convert dtypes, if needed.***

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


-We note the columns match their data type, therefore we don't need to change any dtype

***1.2 We will check the duplicate rows and address them if needed.***

In [5]:
df.duplicated().sum()

0


-We notice that we don't have duplicate rows




***1.3 We will check here for null values and we impute them if needed. (Impute them in a way that prevents data leakage!)***

In [6]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

- We notice here that we have two columns (Item_Weight,Outlet_Size) containing missing values and we will Impute them later

***1.4 We will Check for inconsistent categories and fix them if needed.***

In [7]:
df.select_dtypes('object').value_counts()

Item_Identifier  Item_Fat_Content  Item_Type              Outlet_Identifier  Outlet_Size  Outlet_Location_Type  Outlet_Type      
DRA12            LF                Soft Drinks            OUT035             Small        Tier 2                Supermarket Type1    1
FDV27            Regular           Meat                   OUT019             Small        Tier 1                Grocery Store        1
FDV32            Low Fat           Fruits and Vegetables  OUT035             Small        Tier 2                Supermarket Type1    1
                                                          OUT019             Small        Tier 1                Grocery Store        1
                                                          OUT018             Medium       Tier 3                Supermarket Type2    1
                                                                                                                                    ..
FDJ33            Regular           Snack Foods            OU

In [8]:
df['Item_Fat_Content'].value_counts()

Item_Fat_Content
Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: count, dtype: int64

- We have inconsistent values in the column Item_Fat_Content and we won't fix/replace it

In [9]:
df['Item_Fat_Content'].replace({'LF':'Low Fat','reg':'Regular','low fat':'Low Fat'},inplace=True)

In [10]:
# To confirm
df['Item_Fat_Content'].value_counts()

Item_Fat_Content
Low Fat    5517
Regular    3006
Name: count, dtype: int64

In [11]:
df['Outlet_Type'].value_counts()

Outlet_Type
Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: count, dtype: int64

In [12]:
df['Item_Type'].value_counts()

Item_Type
Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: count, dtype: int64

***1.5 We will check for impossible numeric values and fix them if needed***

In [13]:
df.describe(exclude='object')

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7060.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.643456         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.773750         0.026989    93.826500                1987.000000   
50%      12.600000         0.053931   143.012800                1999.000000   
75%      16.850000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

In [14]:
#We want to check if  Item_Visibility= 0 is an impossible numeric or not
df[df['Item_Visibility']==0.0]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07       19.200          Regular              0.0   
4              NCD19        8.930          Low Fat              0.0   
5              FDP36       10.395          Regular              0.0   
10             FDY07       11.800          Low Fat              0.0   
32             FDP33       18.700          Low Fat              0.0   
...              ...          ...              ...              ...   
8480           FDQ58          NaN          Low Fat              0.0   
8484           DRJ49        6.865          Low Fat              0.0   
8486           FDR20       20.000          Regular              0.0   
8494           NCI54       15.200          Low Fat              0.0   
8500           NCQ42       20.350          Low Fat              0.0   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
5              Baking Goods   51.4008            OUT018   
10    Fruits and Vegetables   45.5402            OUT049   
32              Snack Foods  256.6672            OUT018   
...                     ...       ...               ...   
8480            Snack Foods  154.5340            OUT019   
8484            Soft Drinks  129.9652            OUT013   
8486  Fruits and Vegetables   46.4744            OUT010   
8494              Household  110.4912            OUT017   
8500              Household  125.1678            OUT017   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
5                          2009      Medium               Tier 3   
10                         1999      Medium               Tier 1   
32                         2009      Medium               Tier 3   
...                         ...         ...                  ...   
8480                       1985       Small               Tier 1   
8484                       1987        High               Tier 3   
8486                       1998         NaN               Tier 3   
8494                       2007         NaN               Tier 2   
8500                       2007         NaN               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
5     Supermarket Type2           556.6088  
10    Supermarket Type1          1516.0266  
32    Supermarket Type2          3068.0064  
...                 ...                ...  
8480      Grocery Store           459.4020  
8484  Supermarket Type1          2324.9736  
8486      Grocery Store            45.2744  
8494  Supermarket Type1          1637.8680  
8500  Supermarket Type1          1907.5170  

[526 rows x 12 columns]

- It seems that there is no error in the entry process and that the value 0  of Item_Visibility column is because these products are not available

***2) Identify the features (X) and target (y)***

In [15]:
#We will here separate our data into the feature matrix (X) and the target vector (y)
y=df['Item_Outlet_Sales']
X=df.drop(columns=['Item_Outlet_Sales','Item_Identifier']) #Also, we dropped here the "Item Identifier" feature because it has very high cardinality.
X.head()

Item_Weight Item_Fat_Content  Item_Visibility              Item_Type  \
0         9.30          Low Fat         0.016047                  Dairy   
1         5.92          Regular         0.019278            Soft Drinks   
2        17.50          Low Fat         0.016760                   Meat   
3        19.20          Regular         0.000000  Fruits and Vegetables   
4         8.93          Low Fat         0.000000              Household   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  249.8092            OUT049                       1999      Medium   
1   48.2692            OUT018                       2009      Medium   
2  141.6180            OUT049                       1999      Medium   
3  182.0950            OUT010                       1998         NaN   
4   53.8614            OUT013                       1987        High   

  Outlet_Location_Type        Outlet_Type  
0               Tier 1  Supermarket Type1  
1               Tier 3  Supermarket Type2  
2               Tier 1  Supermarket Type1  
3               Tier 3      Grocery Store  
4               Tier 3  Supermarket Type1

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

##***Pipeline For Preprocessing Numeric Features***##

In [17]:
# Defining lists of types of numeric features
num_cols=X_train.select_dtypes('number').columns
num_cols

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')

In [18]:
# Null values
X_train[num_cols].isna().sum()

Item_Weight                  1107
Item_Visibility                 0
Item_MRP                        0
Outlet_Establishment_Year       0
dtype: int64

In [19]:
X_train[num_cols].describe() #So that we could know what we replace with missing values, whether it is the mean or the median

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year
count  5285.000000      6392.000000  6392.000000                6392.000000
mean     12.904458         0.066007   141.980400                1997.857165
std       4.637034         0.051131    62.629276                   8.392300
min       4.555000         0.000000    31.290000                1985.000000
25%       8.895000         0.027027    94.146200                1987.000000
50%      12.650000         0.054152   144.110200                1999.000000
75%      17.000000         0.094618   186.900300                2004.000000
max      21.350000         0.328391   266.888400                2009.000000

- Since the mean and median are almost equal, we will replace the null values ​​with the mean

In [20]:
# instantiate preprocessors
impute_mean=SimpleImputer(strategy='mean')
scaler = StandardScaler()
scaler
#Instantiate the Pipeline
num_pipe=make_pipeline(impute_mean,scaler)
num_pipe

Pipeline(steps=[('simpleimputer', SimpleImputer()),
                ('standardscaler', StandardScaler())])

###***Pipeline For Preprocessing Ordinal Features***###


In [21]:
# Defining lists of ordinal features
ord_cols=['Outlet_Size','Item_Fat_Content']
# Instantaite and fit the imputer for categorical features
impute_most_freq = SimpleImputer(strategy='most_frequent')
outlet_size_order = ['Small','Medium','High']
Item_Fat_order=['Low Fat','Regular']
ord_encoder =OrdinalEncoder(categories=[outlet_size_order,Item_Fat_order])
scaler=StandardScaler()
# make a pipeline
ord_pipe = make_pipeline(impute_most_freq, ord_encoder, scaler)
ord_pipe

Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='most_frequent')),
                ('ordinalencoder',
                 OrdinalEncoder(categories=[['Small', 'Medium', 'High'],
                                            ['Low Fat', 'Regular']])),
                ('standardscaler', StandardScaler())])

###***Pipeline For Preprocessing Categorical Features***###

In [22]:
# Defining lists of Categorical features
cat_cols = X_train.select_dtypes('object').drop(columns=['Outlet_Size','Item_Fat_Content']).columns
# Prepare Categorical pipeline
impute_missing = SimpleImputer(strategy='constant',fill_value='MISSING')
ohe_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
ohe_pipe = make_pipeline(impute_missing,ohe_encoder)
ohe_encoder

OneHotEncoder(handle_unknown='ignore', sparse_output=False)

In [23]:
# Making a numeric tuple for ColumnTransformer
num_tuple=('numeric',num_pipe,num_cols)
# Making an ordinal_tuple for ColumnTransformer
ord_tuple = ('ordinal', ord_pipe, ord_cols)
# Making a ohe_tuple for ColumnTransformer
ohe_tuple = ('categorical', ohe_pipe, cat_cols)

 Instantiate the ColumnTransformer


In [24]:
# Instantiate with verbose_feature_names_out=False
col_transformer = ColumnTransformer([num_tuple, ord_tuple, ohe_tuple],
                                    verbose_feature_names_out=False)
col_transformer

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('ordinal',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('ordinalencoder',
                                                  OrdinalEncoder(categ...
                                                  StandardScaler())]),
                                 ['Outlet_Size', 'Item_Fat_Content']),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='MISSING',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Type', 'Outlet_Identifier', 'Outlet_Location_Type',
       'Outlet_Type'],
      dtype='object'))],
                  verbose_feature_names_out=False)

***Fit the Transformer on Training Data Only***


In [25]:
# Fit on training data
col_transformer.fit(X_train)

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('ordinal',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('ordinalencoder',
                                                  OrdinalEncoder(categ...
                                                  StandardScaler())]),
                                 ['Outlet_Size', 'Item_Fat_Content']),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='MISSING',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Type', 'Outlet_Identifier', 'Outlet_Location_Type',
       'Outlet_Type'],
      dtype='object'))],
                  verbose_feature_names_out=False)

Transform Training and Test data


In [26]:
# Transform the training data
X_train_processed = col_transformer.transform(X_train)
# Transform the testing data
X_test_processed = col_transformer.transform(X_test)
# View the processed training data
X_train_processed.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
4776     0.817249        -0.712775  1.828109                   1.327849   
7510     0.556340        -1.291052  0.603369                   1.327849   
5828    -0.131512         1.813319  0.244541                   0.136187   
5327    -1.169219        -1.004931 -0.952591                   0.732018   
4810     1.528819        -0.965484 -0.336460                   0.493686   

      Outlet_Size  Item_Fat_Content  Item_Type_Baking Goods  Item_Type_Breads  \
4776     0.287374         -0.740321                     0.0               0.0   
7510     0.287374          1.350766                     0.0               0.0   
5828     0.287374          1.350766                     0.0               0.0   
5327    -1.384048         -0.740321                     1.0               0.0   
4810     0.287374         -0.740321                     0.0               0.0   

      Item_Type_Breakfast  Item_Type_Canned  ...  Outlet_Identifier_OUT045  \
4776                  0.0               0.0  ...                       0.0   
7510                  0.0               0.0  ...                       0.0   
5828                  0.0               0.0  ...                       0.0   
5327                  0.0               0.0  ...                       0.0   
4810                  0.0               0.0  ...                       1.0   

      Outlet_Identifier_OUT046  Outlet_Identifier_OUT049  \
4776                       0.0                       0.0   
7510                       0.0                       0.0   
5828                       0.0                       1.0   
5327                       0.0                       0.0   
4810                       0.0                       0.0   

      Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
4776                          0.0                          0.0   
7510                          0.0                          0.0   
5828                          1.0                          0.0   
5327                          0.0                          1.0   
4810                          0.0                          1.0   

      Outlet_Location_Type_Tier 3  Outlet_Type_Grocery Store  \
4776                          1.0                        0.0   
7510                          1.0                        0.0   
5828                          0.0                        0.0   
5327                          0.0                        0.0   
4810                          0.0                        0.0   

      Outlet_Type_Supermarket Type1  Outlet_Type_Supermarket Type2  \
4776                            0.0                            1.0   
7510                            0.0                            1.0   
5828                            1.0                            0.0   
5327                            1.0                            0.0   
4810                            1.0                            0.0   

      Outlet_Type_Supermarket Type3  
4776                            0.0  
7510                            0.0  
5828                            0.0  
5327                            0.0  
4810                            0.0  

[5 rows x 39 columns]

In [27]:
# View the processed testing data
X_test_processed.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
7503  3.310089e-01        -0.776646 -0.998816                  -1.293807   
2957 -1.179892e+00         0.100317 -1.585194                  -0.102145   
7031  3.784469e-01        -0.482994 -1.595784                   0.136187   
1084  4.213344e-16        -0.415440  0.506592                  -1.532139   
856  -6.426567e-01        -1.047426  0.886725                   0.732018   

      Outlet_Size  Item_Fat_Content  Item_Type_Baking Goods  Item_Type_Breads  \
7503     1.958796         -0.740321                     0.0               0.0   
2957    -1.384048         -0.740321                     0.0               0.0   
7031     0.287374          1.350766                     0.0               0.0   
1084     0.287374          1.350766                     0.0               0.0   
856     -1.384048          1.350766                     0.0               0.0   

      Item_Type_Breakfast  Item_Type_Canned  ...  Outlet_Identifier_OUT045  \
7503                  0.0               0.0  ...                       0.0   
2957                  0.0               0.0  ...                       0.0   
7031                  0.0               1.0  ...                       0.0   
1084                  0.0               0.0  ...                       0.0   
856                   0.0               0.0  ...                       0.0   

      Outlet_Identifier_OUT046  Outlet_Identifier_OUT049  \
7503                       0.0                       0.0   
2957                       1.0                       0.0   
7031                       0.0                       1.0   
1084                       0.0                       0.0   
856                        0.0                       0.0   

      Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
7503                          0.0                          0.0   
2957                          1.0                          0.0   
7031                          1.0                          0.0   
1084                          0.0                          0.0   
856                           0.0                          1.0   

      Outlet_Location_Type_Tier 3  Outlet_Type_Grocery Store  \
7503                          1.0                        0.0   
2957                          0.0                        0.0   
7031                          0.0                        0.0   
1084                          1.0                        0.0   
856                           0.0                        0.0   

      Outlet_Type_Supermarket Type1  Outlet_Type_Supermarket Type2  \
7503                            1.0                            0.0   
2957                            1.0                            0.0   
7031                            1.0                            0.0   
1084                            0.0                            0.0   
856                             1.0                            0.0   

      Outlet_Type_Supermarket Type3  
7503                            0.0  
2957                            0.0  
7031                            0.0  
1084                            1.0  
856                             0.0  

[5 rows x 39 columns]

In [28]:
X_train_processed.describe().round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
count      6392.00          6392.00   6392.00                    6392.00   
mean          0.00            -0.00      0.00                      -0.00   
std           1.00             1.00      1.00                       1.00   
min          -1.98            -1.29     -1.77                      -1.53   
25%          -0.81            -0.76     -0.76                      -1.29   
50%           0.00            -0.23      0.03                       0.14   
75%           0.76             0.56      0.72                       0.73   
max           2.00             5.13      1.99                       1.33   

       Outlet_Size  Item_Fat_Content  Item_Type_Baking Goods  \
count      6392.00           6392.00                 6392.00   
mean         -0.00              0.00                    0.07   
std           1.00              1.00                    0.26   
min          -1.38             -0.74                    0.00   
25%          -1.38             -0.74                    0.00   
50%           0.29             -0.74                    0.00   
75%           0.29              1.35                    0.00   
max           1.96              1.35                    1.00   

       Item_Type_Breads  Item_Type_Breakfast  Item_Type_Canned  ...  \
count           6392.00              6392.00           6392.00  ...   
mean               0.03                 0.01              0.08  ...   
std                0.16                 0.11              0.26  ...   
min                0.00                 0.00              0.00  ...   
25%                0.00                 0.00              0.00  ...   
50%                0.00                 0.00              0.00  ...   
75%                0.00                 0.00              0.00  ...   
max                1.00                 1.00              1.00  ...   

       Outlet_Identifier_OUT045  Outlet_Identifier_OUT046  \
count                   6392.00                   6392.00   
mean                       0.11                      0.11   
std                        0.31                      0.31   
min                        0.00                      0.00   
25%                        0.00                      0.00   
50%                        0.00                      0.00   
75%                        0.00                      0.00   
max                        1.00                      1.00   

       Outlet_Identifier_OUT049  Outlet_Location_Type_Tier 1  \
count                   6392.00                      6392.00   
mean                       0.11                         0.27   
std                        0.31                         0.45   
min                        0.00                         0.00   
25%                        0.00                         0.00   
50%                        0.00                         0.00   
75%                        0.00                         1.00   
max                        1.00                         1.00   

       Outlet_Location_Type_Tier 2  Outlet_Location_Type_Tier 3  \
count                      6392.00                      6392.00   
mean                          0.33                         0.40   
std                           0.47                         0.49   
min                           0.00                         0.00   
25%                           0.00                         0.00   
50%                           0.00                         0.00   
75%                           1.00                         1.00   
max                           1.00                         1.00   

       Outlet_Type_Grocery Store  Outlet_Type_Supermarket Type1  \
count                    6392.00                        6392.00   
mean                        0.12                           0.65   
std                         0.33                           0.48   
min                         0.00                           0.00   
25%                         0.00       

- To make sure the imputation of missing values occurred after the train test split

In [29]:
X_test_processed.isna().sum()

Item_Weight                        0
Item_Visibility                    0
Item_MRP                           0
Outlet_Establishment_Year          0
Outlet_Size                        0
Item_Fat_Content                   0
Item_Type_Baking Goods             0
Item_Type_Breads                   0
Item_Type_Breakfast                0
Item_Type_Canned                   0
Item_Type_Dairy                    0
Item_Type_Frozen Foods             0
Item_Type_Fruits and Vegetables    0
Item_Type_Hard Drinks              0
Item_Type_Health and Hygiene       0
Item_Type_Household                0
Item_Type_Meat                     0
Item_Type_Others                   0
Item_Type_Seafood                  0
Item_Type_Snack Foods              0
Item_Type_Soft Drinks              0
Item_Type_Starchy Foods            0
Outlet_Identifier_OUT010           0
Outlet_Identifier_OUT013           0
Outlet_Identifier_OUT017           0
Outlet_Identifier_OUT018           0
Outlet_Identifier_OUT019           0
O

##***DM Phase 4 - Modeling***##

In [31]:
# Instantiate model
lin_reg = LinearRegression()
#Fit model
lin_reg.fit(X_train_processed, y_train)

LinearRegression()

In [36]:
# Test function with default arguments
evaluate_regression(lin_reg, X_train_processed, y_train, X_test_processed, y_test)


------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 847.126
- MSE = 1,297,559.217
- RMSE = 1,139.105
- R^2 = 0.562

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 804.108
- MSE = 1,194,337.083
- RMSE = 1,092.857
- R^2 = 0.567


-The model performs better on the training data ( R2  0.562 for training data vs. 0.567 for test data)
- R^2_Interpretation:It shows that our model fits the data at a percentage of .55 (not bad, but could be better).
- MAE_Interpretation: On average, our model is incorrect by about 804 dollars.
- MSE_Interpretation: On average, our model is incorrect by about 1194 squared units.
- RMSE_Interpretation: On average, our model is incorrect by about 1092 thousand dollars. Note: This metric penalizes larger errors more than MSE.
We also see here the big difference between MAE and RMSE, meaning that we have large error values ​​and the model penalized it

###***Random Forest model to predict sales.***###

In [39]:
# Instantiate a Default Model
rfreg=RandomForestRegressor(random_state=42)

rf_pipe = make_pipeline(col_transformer, rfreg)
# Fit default model on training data only
rf_pipe.fit(X_train, y_train)

# Evaluate with custom function
evaluate_regression(rf_pipe, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 296.124
- MSE = 182,537.759
- RMSE = 427.244
- R^2 = 0.938

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 766.647
- MSE = 1,218,968.284
- RMSE = 1,104.069
- R^2 = 0.558


-- The results of our model show an overfitting due to a difference between the training and testing data (where r-square = .93 in training versus .55 in testing data)
- MAE_Interpretation: On average, our model is incorrect by about 766 dollars.
- MSE_Interpretation: On average, our model is incorrect by about 1218 squared units.
- RMSE_Interpretation: On average, our model is incorrect by about 1104 thousand dollars. Note: This metric penalizes larger errors more than MSE.
We also see here the big difference between MAE and RMSE, meaning that we have large error values ​​and the model penalized it

***By comparing the results of the two models (Linear Registration and Random Forest), it appears that Linear Registration is better than Random Forest for our model.***

###***GridSearchCV to tune hyperparameters for a Random Forest model***

- First, to explore the parameters available for tuning a random forest regressor.



In [40]:
# Parameters for tuning
rf_pipe.get_params()

{'memory': None,
 'steps': [('columntransformer',
   ColumnTransformer(transformers=[('numeric',
                                    Pipeline(steps=[('simpleimputer',
                                                     SimpleImputer()),
                                                    ('standardscaler',
                                                     StandardScaler())]),
                                    Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
          'Outlet_Establishment_Year'],
         dtype='object')),
                                   ('ordinal',
                                    Pipeline(steps=[('simpleimputer',
                                                     SimpleImputer(strategy='most_frequent')),
                                                    ('ordinalencoder',
                                                     OrdinalEncoder(categ...
                                                     StandardScaler())]),
                           

In [41]:
# Define param grid with options to try
params = {'randomforestregressor__max_depth': [None,10,15,20],
          'randomforestregressor__n_estimators':[10,100,150,200],
          'randomforestregressor__min_samples_leaf':[1,2,3,4],
          'randomforestregressor__max_features':['sqrt','log2',None],
          'randomforestregressor__oob_score':[True,False],
          }

In [43]:
# Instantiate the gridsearch
gridsearch = GridSearchCV(rf_pipe, params, n_jobs=-1, cv = 3, verbose=1)
# Fit the gridsearch on training data
gridsearch.fit(X_train, y_train)
# Obtain best parameters
gridsearch.best_params_

Fitting 3 folds for each of 384 candidates, totalling 1152 fits


{'randomforestregressor__max_depth': 10,
 'randomforestregressor__max_features': None,
 'randomforestregressor__min_samples_leaf': 1,
 'randomforestregressor__n_estimators': 100,
 'randomforestregressor__oob_score': True}

In [44]:
# Define and refit best model
best_rf = gridsearch.best_estimator_
evaluate_regression(best_rf, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 642.386
- MSE = 822,743.553
- RMSE = 907.052
- R^2 = 0.722

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 737.867
- MSE = 1,128,149.455
- RMSE = 1,062.144
- R^2 = 0.591


- The adjusted Random Forest is better than the rest of the models, but it still had an over-fitting the R2 was the best it could be on the test data and close to the training data and the features describe a Y of 0.59.
- The mean absolute error(MAE) is also lower than any other model =737 thousand
- MSE_Interpretation: On average, our model is incorrect by about 1128 squared units.
- RMSE_Interpretation: On average, our model is incorrect by about 1062 thousand dollars.This metric penalizes large errors more than MSE.